SDXL fine tuning

dsocek opened the pull request:
# What does this PR do?

This PR adds fine-tuning for SDXL on Gaudi

libinta commented on examples/stable-diffusion/README.md:
```diff
@@ -294,3 +294,58 @@ image = pipe(prompt, num_inference_steps=50, guidance_scale=7.5).images[0]
+  --use_hpu_graphs \
+  --throughput_warmup_steps 3 \
+  --bf16
+
```
suggest to remove sd related, and you can submit later with just new_version of train_tex_to_image.py

dsocek replied:
Removed as suggested

libinta reviewed (commented):
@dsocek I will provide you a patch with train_text_to_image_sdxl.py in gs system

dsocek commented on examples/stable-diffusion/train_text_to_image.py:
Done

dsocek commented on examples/stable-diffusion/train_text_to_image_sdxl.py:
Done

dsocek commented on examples/stable-diffusion/train_text_to_image_sdxl.py:
```diff
@@ -0,0 +1,1304 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Fine-tuning script for Stable Diffusion XL for text2image."""
```
Done

dsocek commented on examples/stable-diffusion/train_text_to_image_sdxl.py:
```diff
@@ -0,0 +1,1304 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Fine-tuning script for Stable Diffusion XL for text2image."""
+
```
Done

dsocek commented on examples/stable-diffusion/README.md:
```diff
@@ -294,3 +294,58 @@ image = pipe(prompt, num_inference_steps=50, guidance_scale=7.5).images[0]
+  --gaudi_config_name Habana/stable-diffusion \
+  --use_hpu_graphs \
+  --throughput_warmup_steps 3 \
+  --bf16
```
Updated README with it

dsocek commented:
**Additional updates:**
1. Included additional fixes from @libinta (casting in original Gaudi SDXL pipeline)
2. rebased branch with latest OH master
3. added new working SDXL LoRA script for Gaudi
4. updated README with Training example for LoRA and Inference with obtained LoRA weights

libinta commented on examples/stable-diffusion/training/train_text_to_image_lora.py:
```diff
@@ -0,0 +1,1310 @@
+#!/usr/bin/env python
+# coding=utf-8
```
can you add the lora training to train_text_to_image.py with argument?

dsocek replied:
Yes, we can join 2 scripts into One

dsocek replied:
Done

libinta commented on examples/stable-diffusion/README.md:
```diff
@@ -371,3 +371,86 @@ image = pipe(prompt, num_inference_steps=50, guidance_scale=7.5).images[0]
+  --throughput_warmup_steps 3 \
+  --use_hpu_graphs \
+  --bf16
+
```
please align cmd with https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/README_sdxl.md

dsocek replied:
Sounds good

dsocek commented on examples/stable-diffusion/README.md:
```diff
@@ -294,3 +294,31 @@ image = pipe(prompt, num_inference_steps=50, guidance_scale=7.5).images[0]
+  --pretrained_model_name_or_path stabilityai/stable-diffusion-xl-base-1.0 \
+  --pretrained_vae_model_name_or_path stabilityai/sdxl-vae \
+  --dataset_name lambdalabs/pokemon-blip-captions \
+  --resolution 1024 \
```
I think this was typo from original SD.  Stable Diffusion XL was trained at a base resolution of 1024x1024 and SDXL is operating on 1024 as default resolution.  If you look at added LoRA command for training it was fixed to 1024:  https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/README_sdxl.md#training-1

Therefore, I would leave 1024 for fine-tuning SDXL.   If we want to copy the diffusers README command then I can update, but IMO it would not be correct.

dsocek commented on examples/stable-diffusion/README.md:
````diff
@@ -294,3 +294,31 @@ image = pipe(prompt, num_inference_steps=50, guidance_scale=7.5).images[0]
 
 image.save("cat-backpack.png")
 ```
+
````
Done

dsocek commented:
**Updates:**

* Added CI tests
* Refactored readme (split README into inference and training to make it cleaner)
* Merged SDXL fine-tuning and lora into ONE script
* Updated README with changed command lines

regisss commented on examples/stable-diffusion/training/train_text_to_image.py:
```diff
@@ -0,0 +1,1737 @@
+"""
+Fine-tuning script for Stable Diffusion models for text2image with support for LoRA.
+Adapted from the following sources:
+https://github.com/huggingface/diffusers/blob/v0.23.1/examples/text_to_image/train_text_to_image_sdxl.py
+https://github.com/huggingface/diffusers/blob/v0.23.1/examples/text_to_image/train_text_to_image_lora_sdxl.py
```
Maybe this script should be renamed to `train_text_to_image_sdxl.py` as I see that Diffusers also has an example called [`train_text_to_image.py`](https://github.com/huggingface/diffusers/tree/v0.23.1/examples/text_to_image).

libinta replied:
@regisss  should we keep 1 script and use it to train both stable diffusion and stable diffusion xl with parameter?

regisss replied:
Ideally yes, if that doesn't make the script too complicated

regisss reviewed (commented):
@libinta The new training folder for training examples looks good to me :+1:

imangohari1 commented:
@libinta any updates on this?

libinta commented:
Will update next week.
Thanks

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: SM Iman Gohari ***@***.***>
Sent: Friday, February 9, 2024 11:12:59 AM
To: huggingface/optimum-habana ***@***.***>
Cc: Libin Tang ***@***.***>; Mention ***@***.***>
Subject: Re: [huggingface/optimum-habana] Sd sdxl fine tuning (PR #667)


@libinta<https://github.com/libinta> any updates on this?

—
Reply to this email directly, view it on GitHub<https://github.com/huggingface/optimum-habana/pull/667#issuecomment-1936466551>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGLGHZIKWE2RT2USGNFEY7DYSZYLXAVCNFSM6AAAAABCMTBITCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMZWGQ3DMNJVGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

libinta commented:
@dsocek  can you update to the latest?thx

libinta reviewed (commented):
can you update the patch to latest?

dsocek commented:
@libinta I have just rebased to latest OH code

libinta commented:
Daniel,
Can you check one more time to see if it's easier to keep stable diffusion xl and stable diffusion as separate training script? Or combine them?

Thanks,
Libin

From: Daniel Socek ***@***.***>
Sent: Monday, February 12, 2024 3:36 PM
To: huggingface/optimum-habana ***@***.***>
Cc: Libin Tang ***@***.***>; Mention ***@***.***>
Subject: Re: [huggingface/optimum-habana] Sd sdxl fine tuning (PR #667)


@libinta<https://github.com/libinta> I have just rebased to latest OH code

-
Reply to this email directly, view it on GitHub<https://github.com/huggingface/optimum-habana/pull/667#issuecomment-1939785482>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGLGHZNYULHRFS4Y57XGOTDYTKROTAVCNFSM6AAAAABCMTBITCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMZZG44DKNBYGI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

dsocek commented:
> Daniel, Can you check one more time to see if it's easier to keep stable diffusion xl and stable diffusion as separate training script? Or combine them? Thanks, Libin From: Daniel Socek ***@***.***> Sent: Monday, February 12, 2024 3:36 PM To: huggingface/optimum-habana ***@***.***> Cc: Libin Tang ***@***.***>; Mention ***@***.***> Subject: Re: [huggingface/optimum-habana] Sd sdxl fine tuning (PR #667) @libinta<https://github.com/libinta> I have just rebased to latest OH code - Reply to this email directly, view it on GitHub<[#667 (comment)](https://github.com/huggingface/optimum-habana/pull/667#issuecomment-1939785482)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGLGHZNYULHRFS4Y57XGOTDYTKROTAVCNFSM6AAAAABCMTBITCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMZZG44DKNBYGI>. You are receiving this because you were mentioned.Message ID: ***@***.***>

@libinta   I think better would be to have separate train scripts for SDXL and SD.  This would be easier to implement and also easier to maintain IMO.  It also is more similar to how diffusers are arranged.   So we would then have 2 scripts but each could cover multiple fine tunings approaches.

libinta commented on examples/stable-diffusion/training/README.md:
```diff
@@ -0,0 +1,207 @@
+pipe = GaudiStableDiffusionXLPipeline.from_pretrained(
+  model_id,
+  scheduler=GaudiEulerDiscreteScheduler.from_pretrained(model_id, subfolder="scheduler"),
+  torch_dtype=torch.bfloat16,
```
can you change GaudiEulerDiscreteScheduler to DDPMScheduler like original script as we observed noise image with EulerDiscreteScheduler ?

dsocek replied:
For this change, do we need to implement GaudiDDDPMScheduler?

regisss replied:
I guess yes. Maybe it works well with `GaudiDDIMScheduler`?

dsocek replied:
@regisss  @libinta  we tested now with updated diffusers Euler Discrete scheduler works good (no noise images any more observed)

libinta reviewed (commented):
@dsocek As we discussed, can you rename the training script to train_text_to_image_sdxl.py ?

atakaha commented:
@libinta and @dsocek 
There is diffusers  version dependency issue between driver 1.13 and 1.14. 
driver 1.13 use diffusers 0.23.1 and  driver 1.14 use diffusers 0.26.3. This 0.23.1 has "text_encoder_lora_state_dict" but 0.26.3 not. Do we support only 1.13 or 1.14 or both? I recommend go for 1.14.

libinta commented:
Yes, please update to the latest one.thanks.

From: Akihiro Takahashi ***@***.***>
Sent: Tuesday, February 13, 2024 2:14 PM
To: huggingface/optimum-habana ***@***.***>
Cc: Libin Tang ***@***.***>; Mention ***@***.***>
Subject: Re: [huggingface/optimum-habana] Sd sdxl fine tuning (PR #667)


@libinta<https://github.com/libinta> and @dsocek<https://github.com/dsocek>
There is diffusers version dependency issue between driver 1.13 and 1.14.
driver 1.13 use diffusers 0.23.1 and driver 1.14 use diffusers 0.26.3. This 0.23.1 has "text_encoder_lora_state_dict" but 0.26.3 not. Do we support only 1.13 or 1.14 or both? I recommend go for 1.14.

-
Reply to this email directly, view it on GitHub<https://github.com/huggingface/optimum-habana/pull/667#issuecomment-1942728889>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGLGHZOBK5FN4CXGWZRKS6LYTPQTVAVCNFSM6AAAAABCMTBITCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNBSG4ZDQOBYHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

dsocek commented:
> @dsocek As we discussed, can you rename the training script to train_text_to_image_sdxl.py ?

@libinta  renamed

dsocek commented:
@libinta
- We validated on 1.14 and latest diffusers
- We updated to support PEFT as needed
- We also updated tests
- I squashed all commits into single commit with 4 co-authors for easier merge
- I moved this PR from draft to actual PR
- Scheduler can be updated in future commit (if needed)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_667). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

regisss commented on examples/stable-diffusion/training/README.md:
````diff
@@ -0,0 +1,214 @@
+  --bf16
+```
+
+**Notes:**
+
+* SDXL's VAE is known to suffer from numerical instability issues. This is why we
+also expose a CLI argument namely `--pretrained_vae_model_name_or_path` that lets
+you specify the location of a better VAE (such as this one).
````
```suggestion
> [!NOTE] 
> SDXL's VAE is known to suffer from numerical instability issues. This is why we also expose a CLI argument namely `--pretrained_vae_model_name_or_path` that lets you specify the location of a better VAE (such as this one).
```

imangohari1 replied:
Will push the changes.

regisss commented on examples/stable-diffusion/training/train_text_to_image_sdxl.py:
```diff
@@ -0,0 +1,1695 @@
+    import wandb
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.23.1")
```
```suggestion
check_min_version("0.26.0")
```
Could you also check the changes in this script between v0.23 and v0.26 and integrate them if they make sense?

imangohari1 replied:
The changes are here: https://github.com/huggingface/diffusers/commit/c2717317f03b12535cfd02b477ace61189e10e4b
and we've looked at them to integrate the PEFT, otherwise 0.23 does NOT work. 
Just changed this line to 0.26.0 and it works fine. 
Will push the change.

libinta replied:
@imangohari1 can you update this? we want to close this PR.

imangohari1 replied:
@libinta 
We pushed the changes. 
Feel free to review and move forward.

imangohari1 commented on examples/stable-diffusion/training/requirements.txt:
```diff
@@ -0,0 +1 @@
+peft==0.7.0
```
Just tested it now and it does. 
Although the version was taken from here: https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/requirements_sdxl.txt#L8

regisss reviewed (approved):
LGTM!
